Improved Composer Support

thejamescollins opened the pull request:
Related to wcz/pr510

Makes composer install support more generic. Now a `composer install` is run for any checkout that contains a `composer.json` file.  Previously `composer install` was only run for WooCommerce directories only.

Also adds support for WooCommerce 6.0+'s monorepo structure, where the composer.json file is in a `plugins/woocommerce` subdirectory.

These changes allow our CI's to support the upcoming WooCommerce v6, as well as the upcoming WooCommerce Subscriptions v4 (which has composer dependencies).

This PR also removes the need to revert to composer v1 and back again when installing WooCommerce dependencies.

thejamescollins committed:
Add Support for WooCommerce 6.0+ monorepo structure

Where the composer dependencies are in plugins/woocommerce/composer.json rather than in the root of the git repository.

thejamescollins committed:
Fix  [InvalidArgumentException]                    Version "2>/dev/null" could not be found.

thejamescollins committed:
Install composer dependencies for any dir that contains composer.json with an uncommited vendor dir

Ensures WC dependencies are installed (both in 5.x and 6.0 structure).
Ensures WCS dependencies are installed.
Ensures WCB dependencies are NOT installed.

thejamescollins committed:
Retry a failed composer install with --ignore-platform-reqs

Simpler than downgrading composer to v1 in order to be able to install old WC versions.

the-csaba committed:
CS formatting

the-csaba commented on bin/gvs:
```diff
@@ -95,30 +95,53 @@ for REPO in "${REPOSITORIES[@]}"; do
-        echo "Switching composer version to install WooCommerce dependencies"
-        composer_self_update rollback
+    # Directories to look for a composer.json file in
+    declare -a dirs=("$REPO_PATH" "$REPO_PATH/plugins/woocommerce")
```
```suggestion
    DIRS=("$REPO_PATH" "$REPO_PATH/plugins/woocommerce")
```

The `declare` part is not necessary.

Please use all caps for variable names to match existing code (need more editing; I did not update everywhere).

the-csaba commented on bin/gvs:
```diff
@@ -95,30 +95,53 @@ for REPO in "${REPOSITORIES[@]}"; do
-        composer_self_update rollback
+    # Directories to look for a composer.json file in
+    declare -a dirs=("$REPO_PATH" "$REPO_PATH/plugins/woocommerce")
+    for dir in "${dirs[@]}"; do
```
```suggestion
    for DIR in "${DIRS[@]}"; do
```

the-csaba commented on bin/gvs:
```diff
@@ -40,7 +40,7 @@ environment_info() {
 }
 
 composer_self_update() {
```
function `composer_self_update` no longer used. can be deleted (line 42-47)

the-csaba commented on bin/gvs:
```diff
@@ -95,30 +95,53 @@ for REPO in "${REPOSITORIES[@]}"; do
+
+        # Clean up files and then install composer dependencies
+
+        if [ -f "$dir/composer.lock" ] && [ $(git ls-files composer.lock | wc -l) -eq "0" ]; then
+          $VERBOSE_OUTPUT && echo "$dir/composer.lock is not committed to git. Deleting."
+          rm -f "$dir/composer.lock"
+        fi
```
This logic repeated multiple times, please extract to a function

Also please put `$(...)` structures inside double qoute

the-csaba commented on bin/gvs:
```diff
@@ -95,30 +95,53 @@ for REPO in "${REPOSITORIES[@]}"; do
+          rm -f "$dir/composer.lock"
+        fi
+
+        if [ -d "$dir/plugins" ] && [ $(git status --porcelain | grep plugins/ | wc -l) -ne "0" ]; then
```
Is a difference between using `git status` and `git ls`? it looks to me doing the same thing, so can be unified.

thejamescollins replied:
Good idea, we can achieve the same result with either.
I've standardised this in 67861683932c30a9b5f5900019433c0f836050f2.

the-csaba commented on bin/gvs:
```diff
@@ -95,30 +95,53 @@ for REPO in "${REPOSITORIES[@]}"; do
+        fi
+        if [ -d "$dir/vendor" ]; then
+          $VERBOSE_OUTPUT && echo "$dir/vendor exists"
+          if [ $(git ls-files vendor | wc -l) -ne "0" ]; then
+            $VERBOSE_OUTPUT && echo "$dir/vendor is committed to git, so no composer install required. Skipping."
+            break
```
Should not be this the first thing to check?

Also can the deletion logic can be combine to the rest (see comment on lines +110 to +113).

thejamescollins replied:
I think in this instance even if `/vendor/` exists and is committed, there still may be cleanup required for an uncommitted `composer.lock` file.

Use case: switching from an older WooCommerce such as 4.9.0 (which [has vendor and composer.lock both committed](https://github.com/woocommerce/woocommerce/tree/4.9.0)), to a new version such as 5.9.0, which has neither committed. In that case no composer install is required, but composer.lock should also be deleted and not left on the filesystem.

So my vote is to leave this code section as-is.

the-csaba commented on bin/gvs:
```diff
@@ -95,30 +95,53 @@ for REPO in "${REPOSITORIES[@]}"; do
+    for dir in "${dirs[@]}"; do
+      if [ -f "$dir/composer.json" ]; then
+        $VERBOSE_OUTPUT && echo "composer.json exists at $dir/composer.json"
+        cd "$dir"
```
```suggestion
        cd "$dir" || exit 1
```

the-csaba commented on bin/gvs:
```diff
@@ -95,30 +95,53 @@ for REPO in "${REPOSITORIES[@]}"; do
+          $VERBOSE_OUTPUT && echo "Composer install failed. Retrying with --ignore-platform-reqs."
+          composer install --no-suggest --no-dev --quiet --ignore-platform-reqs
+        fi
+        cd "$REPO_PATH"
```
```suggestion
```

This can be disappear, (see line 88).

the-csaba reviewed (changes requested):
Hi @om4james 

Thank you, I like it very much!

I had minor requests. Please look at the individual comments.

thejamescollins committed:
Apply some suggestions from @om4csaba's code review

Co-authored-by: om4csaba <45931795+om4csaba@users.noreply.github.com>

thejamescollins committed:
Remove unused function

thejamescollins committed:
Update variable name as a result of b0dbb87921ec86335edf2f7e36fb6c0a23d269b7

thejamescollins committed:
Extract to function

Suggested in https://github.com/OM4/git-version-switcher/pull/2/files#r755741523

om4james commented:
TODO: 2461260...b0dbb87  is somehow causing faililng PHP 7.2 tests sometimes?

https://app.circleci.com/pipelines/github/OM4/woocommerce-zapier/1234/workflows/1b07cff6-dc09-46d4-bb30-
https://app.circleci.com/pipelines/github/OM4/woocommerce-zapier/1234/workflows/1b07cff6-dc09-46d4-bb30-ddfd44a01be5/jobs/5701

```
composer install --optimize-autoloader $IGNORE_OPTION
  - Updating om4/git-version-switcher dev-feature/woocommerce-monorepo (2461260 => b0dbb87):  Checking out b0dbb87921
```

thejamescollins committed:
Standardise the usage of `git status` rather than `git ls-files`

Thanks to @om4csaba for the suggestion.

thejamescollins committed:
Fix for failling PHP 7.3 Matrix Tests with old WooCommerce versions

https://app.circleci.com/pipelines/github/OM4/woocommerce-zapier/1295/workflows/dede4db1-225c-4f20-97f6-6b155dfe9adf/jobs/5885

thejamescollins committed:
Allow composer 2.2 to install dependencies for older WooCommerce versions

om4james commented:
Hi @om4csaba I think I've addressed all of your feedback here, thank you!

bd915df...53e04bb

The above diff includes some other changes/improvements in order to get https://github.com/OM4/woocommerce-zapier/compare/c189c51ec246cd3700dc38c8b494747e1e3de9a2...cca7233418029e255a6a93fc0f19e29cc0f6dc6d working.

Over to you (when you have time). Thanks

the-csaba committed:
Better repo cleanup

- Use git built-in functionalities
- move composer install to own function
- unify verbose-only echo

Co-authored-by: James Collins <james@om4.com.au>

om4james commented:
Thank you @om4csaba 💯

om4csaba commented:
Hi @om4james 

It was your hard work created the base for that.

Thanks!
